core/src/main/scala/org/apache/spark/h2o/converters/ReadConverterContext.scala

```diff
@@ -52,23 +54,43 @@ trait ReadConverterContext {
   def getDouble(columnNum: Int): Option[Double]
   def getString(columnNum: Int): Option[String]
 
-  private type Getter = Int => Option[Any]
-
-  val get: DataType => Getter = Map[DataType, Getter](
-    ByteType -> getByte,
-    ShortType -> getShort,
-    IntegerType -> getInt,
-    LongType -> getLong,
-    FloatType -> getFloat,
-    DoubleType -> getDouble,
-    BooleanType -> getBoolean,
-    StringType -> getUTF8String,
-    TimestampType -> getTimestamp
-  ) withDefault (t => throw new scala.IllegalArgumentException(s"Type $t not supported for conversion from H2OFrame to Spark's Dataframe"))
-
-  def getNullable(dataType: DataType): (Int => Any) = (i: Int) => get(dataType)(i) orNull
-
   def numRows: Int
   def increaseRowIdx() = rowIdx += 1
+
   def hasNext = rowIdx < numRows
+
+  case class OptionReader(name: Any, apply: Int => Option[Any]) {
+  }
+
+//  type OptionReader = Int => Option[Any]
+
+  import SupportedTypes._
+
+  val ReaderPerType = Map[SupportedType, Int => Option[Any]](
+    Byte -> getByte,
+    Short -> getShort,
+    Integer -> getInt,
+    Long -> getLong,
+    Float -> getFloat,
+    Double -> getDouble,
+    Boolean -> getBoolean,
+    String -> getUTF8String,
+    UTF8 -> getUTF8String,
+    Timestamp -> getTimestamp
+  ) withDefault (t => throw new scala.IllegalArgumentException(s"Type $t not supported for conversion from H2OFrame to Spark's Dataframe"))
+
+  def readerFor(dt: DataType): OptionReader = OptionReader(bySparkType(dt), ReaderPerType(bySparkType(dt)))
+
+  case class Reader(name: Any, apply: Int => Any)
+
+  val readerMapByName: Map[NameOfType, Reader]
```

Comment:
Yes, it was. Now reorganized, that last part was missing. Now it's shared, since it's the same functionality.